File: src/main/java/no/bekk/java/examples/ScopeAndClosures.java
package no.bekk.java.examples;

public class ScopeAndClosures {
    @FunctionalInterface
    interface Function0 {
        public void apply();
    }

    private int num = 1;

    void closure() {
        int num = 2;
        Function0 fun = () -> {
            this.num = 3; // ok
            //num = 4; //Causes compile time error
            System.out.println(this.num); // prints 3
            System.out.println(num); // prints 2
        }; fun.apply();}

    void scope() {
        int num = 2;
        Function0 fun = () -> {
            //int num = 3; //Causes compile time error
        }; fun.apply();}
}

File: src/main/scala/no/bekk/scala/examples/ScopeAndClosures.scala
package no.bekk.scala.examples

class ScopeAndClosures {
  var num = 1

  def closure {
    var num = 2
    (() => {
      this.num = 4 // ok
      num = 3 // ok
      println(this.num) // prints 4
      println(num) // prints 3
    })()
  }

  def scope {
    var num = 2
    (() => {
      var num = 3 // shadowing of `num` in parent scope
      println(num) // prints 3
    })()
  }
}
